.github/workflows/windows.yml

```diff
@@ -42,52 +42,204 @@ env:
+          # Job concurrency ensures that runners never get assigned twice.
+          reserved_since=$(date +\%s)
+          gh api "$self_hosted_runner_scope/$(cat $runner_id)/labels" \
+            -f "labels[]=reserved-for:$unique_id" \
+            -f "labels[]=reserved-since:$reserved_since" \
+            --method POST --silent
+          echo "selected_runner_label=reserved-for:$unique_id" | tee -a $GITHUB_OUTPUT
+          echo 'is_self_hosted=true' | tee -a $GITHUB_OUTPUT
+
+  # In the unlikely event a self-hosted runner was selected and reserved but it
+  # goes down before the workload starts, cancel the workflow run.
+  runner-timeout:
+    needs:
+      - runner-select
+    if: ${{ fromJSON(needs.runner-select.outputs.is-self-hosted) }}
+    name: Runner timeout
+    runs-on: ubuntu-latest
+    steps:
+      - name: Wait a bit
+        run: sleep 30
+
+      - name: Cancel if workload job is still queued
+        run: |
+          run_url=/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}
+          export GH_TOKEN=${{ secrets.GITHUB_TOKEN }}
+
+          if [ "$(gh api "$run_url/jobs" \
+            | jq -er --arg id '${{ needs.runner-select.outputs.unique-id }}' \
+              '.jobs[] | select(.name | contains("[" + $id + "]")) | .status'
+          )" = queued ]; then
+            echo 'Timeout waiting for runner assignment!'
+            echo 'Hint: does this repo have permission to access the runner group?'
+            echo 'Hint: https://github.com/organizations/servo/settings/actions/runner-groups'
+            echo
+            echo 'Cancelling workflow run'
+            gh api "$run_url/cancel" --method POST
+            exit 1
+          fi
+
   build:
-    name: Windows Build
-    runs-on: windows-2022
+    needs:
+      - runner-select
+    name: Windows Build [${{ needs.runner-select.outputs.unique-id }}]
+    runs-on: ${{ needs.runner-select.outputs.selected-runner-label }}
     steps:
-      - uses: actions/checkout@v4
-        if: github.event_name != 'pull_request_target'
-      # This is necessary to checkout the pull request if this run was triggered via a
-      # `pull_request_target` event.
-      - uses: actions/checkout@v4
-        if: github.event_name == 'pull_request_target'
+      - if: ${{ ! fromJSON(needs.runner-select.outputs.is-self-hosted) && github.event_name != 'pull_request_target' }}
+        uses: actions/checkout@v4
+      - if: ${{ ! fromJSON(needs.runner-select.outputs.is-self-hosted) && github.event_name == 'pull_request_target' }}
+        uses: actions/checkout@v4
         with:
           ref: ${{ github.event.pull_request.head.sha }}
           fetch-depth: 2
+
+      # Faster checkout for self-hosted runner that uses prebaked repo.
+      - if: ${{ fromJSON(needs.runner-select.outputs.is-self-hosted) && github.event_name != 'pull_request_target' }}
+        run: git fetch --depth=2 origin $env:GITHUB_SHA
```

Comment:
I'm not sure that `--depth 2` is necessary any longer now that linting has its own job. Maybe @sagudev can confirm.

Reply:
Indeed, `--depth 1` should be enough.

Reply:
done in 4f21cac163c17370bb14a5882150dadd2e678e08

Reply:
done in 4f21cac163c17370bb14a5882150dadd2e678e08